### PR TITLE
- Events DB: User

### LINKED ISSUE #778: Event db: User table
We need a users table for the event database. This gives foreign keys in several other tables. [WriteMarker](https://github.com/0chain/0chain/issues/768) [ReadMarker](https://github.com/0chain/0chain/issues/764) [Allocation](https://github.com/0chain/0chain/issues/763) [Curator](https://github.com/0chain/0chain/issues/762) and [Allocation Pool](https://github.com/0chain/0chain/issues/760).

The state object is typically handelled outside smart contracts. The MPT data object is [State](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/chaincore/state/state.go#L16).

The user is part of the key the user state is stored under. [getState](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/chaincore/chain/state.go#L441).
Nonce will be added as part of https://github.com/0chain/0chain/issues/765.

Change balance events will need to be emitted each time a user's wallet balance changes. Changes to:
-  [AddTransfer](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/chaincore/chain/state/state_context.go#L128) 
-  [AddMint](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/chaincore/chain/state/state_context.go#L145).
- others ...?

| Users         | type   | State         | type    | notes                                                |
|---------------|--------|---------------|---------|------------------------------------------------------|
| TransactionId | uint   | State.TxHash  | string  | foreign key to last transaction to change user state |
| Balance       | int64  | State.Balance | Balance |                                                      |
| UserId        | string |               | string  | used in MPT key                                      |
| Nonce         | int64  |               | int64   | added issue 765                                      |

### LINKED ISSUE #933: Blocker on UserTable
to mark UserTable: Eventdb #778 to completion, we need to figure out how to emit event from:
- [PutClient](https://github.com/0chain/0chain/blob/6af8438d63636e08e1dd19f41ccc4143c4fd4a3b/code/go/0chain.net/chaincore/client/entity.go#L291)
- [GetClient](https://github.com/0chain/0chain/blob/6af8438d63636e08e1dd19f41ccc4143c4fd4a3b/code/go/0chain.net/chaincore/client/entity.go#L272)

reference PR: #864

### PR DESCRIPTION
**Pull request by ssardana96**
Tasks to complete before merging PR:
- [ ] All places where user balance is updated to emit event there
- [ ] Add client
- [ ] Unit tests for all scenarios to be added/tested
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master BEFORE merging this PR  
  
Associated PRs (Link as appropriate):
...

### CODE REVIEW

File: code/go/0chain.net/chaincore/chain/state/state_context.go [hunk 1]
```diff
@@ -133,6 +134,10 @@ func (sc *StateContext) AddTransfer(t *state.Transfer) error {
 		return state.ErrInvalidTransfer
 	}
 	sc.transfers = append(sc.transfers, t)
+	err := sc.emitHandleAddTransfer(t)
```
Comment (kushthedude) [inline code/go/0chain.net/chaincore/chain/state/state_context.go]: Is the `handle` prefix necessary?

File: code/go/0chain.net/smartcontract/dbs/event/eventdb.go [hunk 2]
```diff
@@ -6,7 +6,10 @@ import (
 	"time"
 )
 
-const DefaultQueryTimeout = 5 * time.Second
+const (
+	DefaultQueryTimeout = 5 * time.Second
+	ErrNoEventDb        = "event db is not initialized"
```
Comment (kushthedude) [inline code/go/0chain.net/smartcontract/dbs/event/eventdb.go]:
```suggestion
	ErrNoEventDb        = Errors.new("event db is not initialized")
```

File: code/go/0chain.net/smartcontract/dbs/event/user.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+type User struct {
+	gorm.Model
+	TransactionID string        `json:"transaction_id"`
+	Balance       state.Balance `json:"balance"`
+	UserID        string        `json:"user_id"`
+
+	// Non sql fields
+	Amount state.Balance `gorm:"-"`
```
Comment (kushthedude) [inline code/go/0chain.net/smartcontract/dbs/event/user.go:17]: Is it being stored as array? Not familiar with the `-` notation in the gorm what does it mean, will it not create a column for this?

**Review by kushthedude — COMMENTED** [review]
Some neats

File: code/go/0chain.net/smartcontract/dbs/event/user.go (hunk 3 above)
Comment (ssardana96) [reply code/go/0chain.net/smartcontract/dbs/event/user.go:17]: Yes

**Comment by peterlimg** [comment]
Replaced by #1234

**Pull request closed without merging**